### ISSUE TITLE
FP8 grouped GEMM returns output with uninitialized regions → non-deterministic NaN in MoE training

### ISSUE THREAD

**Issue by HuangWei-95** [AMD-AGI/Primus-Turbo#397]
## Summary

The FP8 grouped GEMM (TRITON backend) returns an output tensor whose **uncovered
regions are never written** and come back as uninitialized memory. In MoE
training this surfaces as the input/weight gradient containing `NaN`
(`grad norm = nan` from the first step), non-deterministically — it depends on
what the caching allocator hands back.

## Environment

- primus_turbo `0.3.0+3c39ef2` (also present on `main`)
- transformer_engine `2.12.0.dev0+40434cf6`, aiter `0.1.12.post2.dev214`, torch `2.10`
- AMD Instinct **MI355X (gfx950)**, ROCm
- Workload: GPT-OSS-20B MoE, FP8 `e4m3` tensorwise, 32 local experts (EP=1),
  `PRIMUS_TURBO_GROUPED_GEMM_BACKEND=TRITON`.

## Root cause

`primus_turbo/triton/grouped_gemm/grouped_gemm_fp8_kernel.py`:

- The output is allocated uninitialized — `out = torch.empty((M_total, N), ...)`
  (and the `[G, OUT_M, OUT_N]` variable-K variants).
- The persistent kernel iterates tiles per group using `group_offs`/`group_lens`
  and stores with mask `rm_s < M_g` at base `group_offs[g]`, i.e. it only writes
  rows `[0, sum(group_lens))`.

So these output elements are never written and stay uninitialized:

1. **forward / dgrad** `[M_total, N]`: when activations are FP8-padded so
   `M_total > sum(group_lens)`, the trailing rows are never written.
2. **wgrad (variable-K)** `[G, OUT_M, OUT_N]`: empty-expert blocks
   (`group_len == 0`, common in real routing) and per-group `OUT_M` padding.

Under training memory churn those bytes are frequently `NaN`/`Inf`, which then
propagates into the gradients. Existing tests never catch it because they always
use `M_total == sum(group_lens)` with no empty experts.

## Deterministic reproduction

Pure `torch` + `primus_turbo` (no training stack). It poisons the caching
allocator with finite sentinels, frees them, runs the op, and counts
returned-output elements that still equal a sentinel (= memory the op never
wrote). GEMM outputs here are ~1e-5, so they can never equal the sentinels —
every hit is genuinely uninitialized memory.

Observed (real GPT-OSS-20B shapes, `G=32, K=N=2880, sum=65536, M_total=65760`,
6 empty experts):

```
dgrad  (65760, 2880)    uninitialized elems ≈ 94,000
wgrad  (32, 2880, 2880)  uninitialized elems ≈ 137,000
```

<details><summary>repro_grouped_gemm_fp8_uninit.py</summary>

```python
import torch

from primus_turbo.pytorch.core.backend import BackendType, GlobalBackendManager
from primus_turbo.pytorch.core.low_precision import ScalingGranularity, float8_e4m3
from primus_turbo.pytorch.core.quantized_tensor import QuantizedTensor
from primus_turbo.pytorch.kernels.grouped_gemm.grouped_gemm_fp8_impl import (
    grouped_gemm_fp8_impl,
    grouped_gemm_fp8_variable_k_impl,
)
from primus_turbo.pytorch.kernels.grouped_gemm.grouped_gemm_utils import (
    group_offs_from_lens,
)

DEV = "cuda"
DT = torch.bfloat16
GRAN = ScalingGranularity.TENSORWISE
SENTINELS = (7.0, 123.0)

G, K, N = 32, 2880, 2880
REAL_LENS = [372, 0, 3145, 34, 0, 2816, 1677, 2108, 2, 5264, 1148, 1281, 37, 0,
             1463, 545, 8, 7822, 540, 0, 0, 6590, 7876, 5426, 3305, 7269, 18, 0,
             2369, 4309, 24, 88]
PAD = 224


def q(x, axis=-1):
    return QuantizedTensor.quantize(x, float8_e4m3, GRAN, axis=axis)


def poison(shape):
    for v in SENTINELS:
        blocks = [torch.full(shape, v, device=DEV, dtype=DT) for _ in range(4)]
        for b in blocks:
            b.add_(0.0)
        del blocks


def count_stale(out):
    of = out.float()
    return sum(int((of == v).sum()) for v in SENTINELS)


def main():
    GlobalBackendManager.set_grouped_gemm_backend(BackendType.TRITON)
    gl = torch.tensor(REAL_LENS, dtype=torch.int64, device=DEV)
    go = group_offs_from_lens(gl)
    S = int(gl.sum())
    M = S + PAD
    print(f"G={G} K={K} N={N} sum={S} M_total={M} pad={PAD} empty={int((gl==0).sum())}")

    a = q(torch.randn(M, K, device=DEV, dtype=DT))
    gout = q(torch.randn(M, N, device=DEV, dtype=DT))
    b = q(torch.randn(G, N, K, device=DEV, dtype=DT))

    poison((M, K))
    dgrad = grouped_gemm_fp8_impl(
        gout.qdata, b.qdata, gout.scale_inv, b.scale_inv, gl, go,
        trans_a=False, trans_b=False, out_dtype=DT, granularity=GRAN.value,
        num_cu=None, default_backend=BackendType.TRITON.value)
    torch.cuda.synchronize()
    print("dgrad uninitialized elems =", count_stale(dgrad))

    poison((G, K, N))
    wgrad = grouped_gemm_fp8_variable_k_impl(
        a.qdata, gout.qdata, a.scale_inv, gout.scale_inv, gl, go,
        trans_a=True, trans_b=False, trans_c=True, out_dtype=DT,
        granularity=GRAN.value, num_cu=None, default_backend=BackendType.TRITON.value)
    torch.cuda.synchronize()
    print("wgrad uninitialized elems =", count_stale(wgrad))


if __name__ == "__main__":
    main()
```

</details>

It was also confirmed end-to-end: hooking `FP8GroupedGemmTensorFunc.backward` on
the first NaN in real training showed all *inputs* clean (`grad_out` amax
~1.5e-4, scales finite) while the returned gradient contained NaN; the only
difference from a clean run is allocator contents.

## Note: the fix is not just `empty -> zeros`

Changing the kernel's `torch.empty` to `torch.zeros` is **not sufficient** on
this stack: with the kernel allocating `torch.zeros` (or even an explicit
`out.fill_(...)`), the tensor *returned by the `grouped_gemm_fp8_impl` custom op*
still exposes uninitialized regions in the repro. This suggests the returned
tensor is not (entirely) the kernel's python-allocated `out` — likely an
interaction with the `torch.library.custom_op` wrapper / functionalization,
and/or the kernel leaving scattered write-holes beyond the obvious padding rows.

A correct fix should guarantee the **returned** output is fully defined, e.g.:
- have the triton kernel explicitly store zeros for every uncovered tile
  (padding rows, empty-expert blocks, partial-tile remainder), or
- zero the uncovered region of the returned tensor at the op level.

## Workaround

Route MoE grouped GEMM through Transformer Engine
(`use_turbo_grouped_gemm: false` on the Primus side). It fully writes its output
and avoids the NaN (and in our setup also lowered peak VRAM ~98% -> ~76%).

**Comment by kyle-256** [comment]
https://github.com/AMD-AGI/Primus-Turbo/pull/408/commits
